src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,443 @@
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.SemanticModelReuse;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+using VsWebSite;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    using LspDocumentSymbol = DocumentSymbol;
+
+    /// <summary>
+    /// Responsible for updating data related to Document outline. It is expected that all public methods on this type
+    /// do not need to be on the UI thread. Two properties: <see cref="SortOption"/> and <see cref="SearchText"/> are
+    /// intended to be bound to a WPF view and should only be set from the UI thread.
+    /// </summary>
+    internal sealed partial class DocumentOutlineViewModel : INotifyPropertyChanged, IDisposable
+    {
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+        private readonly ITaggerEventSource _taggerEventSource;
+        private readonly ITextView _textView;
+        private readonly ITextBuffer _textBuffer;
+        private readonly IThreadingContext _threadingContext;
+
+        /// <summary>
+        /// Queue that computes the new model and updates the UI state.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _workQueue;
+
+        /// <summary>
+        /// Queue responsible for updating the ui after a change/move happens.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _selectionQueue;
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        // Mutable state.  Should only update on UI thread.
+
+        private SortOption _sortOption_doNotAccessDirectly = SortOption.Location;
+        private string _searchText_doNotAccessDirectly = "";
+        private ImmutableArray<DocumentSymbolDataViewModel> _documentSymbolViewModelItems_doNotAccessDirectly = ImmutableArray<DocumentSymbolDataViewModel>.Empty;
+        private DocumentOutlineViewState _lastPresentedViewState_doNotAccessDirectly;
+
+        /// <summary>
+        /// Use to prevent reeentrancy on navigation/selection.
+        /// </summary>
+        private bool _isNavigating_doNotAccessDirectly;
+
+        public DocumentOutlineViewModel(
+            ITextView textView,
+            IThreadingContext threadingContext,
+            ILanguageServiceBroker2 languageServiceBroker,
+            IAsynchronousOperationListener asyncListener)
+        {
+            _languageServiceBroker = languageServiceBroker;
+            _textView = textView;
```

Comment:
This is fishy that the ITextView is being held onto here, since the active view was requested when this was created. If that was the secondary view in a split case, and the split is removed, this view will be closed. I'd expect:

1. This type is given the VsCodeWindowViewTracker and that's held onto instead.
2. The fetching of the text buffer is calls GetActiveView().TextBuffer, since text buffer is shared across the views. (feel free to add a method to VsCodeWindowViewTracker to do this to make it clear it's safe).
3. The place you are fetching the caret, call GetActiveView() instead.

Reply:
indeed.  good catch.  Have revised this as per your design.  Thanks!